google-cloud-firestore/src/main/java/com/google/cloud/firestore/TransactionOptions.java

```diff
@@ -83,10 +179,264 @@ public static TransactionOptions create(@Nonnull Executor executor) {
    * @param executor The executor to run the user callback code on.
    * @param numberOfAttempts The number of execution attempts.
    * @return The TransactionOptions object.
+   * @deprecated as of 2.0.0, replaced by {@link ReadWriteOptionsBuilder#setExecutor(Executor)} and
+   *     {@link ReadWriteOptionsBuilder#setNumberOfAttempts(int)}
+   * @see #createReadWriteOptionsBuilder()
+   */
+  @Nonnull
+  @Deprecated
+  public static TransactionOptions create(@Nullable Executor executor, int numberOfAttempts) {
+    return createReadWriteOptionsBuilder()
+        .setExecutor(executor)
+        .setNumberOfAttempts(numberOfAttempts)
+        .build();
+  }
+
+  /**
+   * @return a new builder with default values applicable to configuring options for a read-write
+   *     transaction.
+   */
+  @Nonnull
+  public static ReadWriteOptionsBuilder createReadWriteOptionsBuilder() {
+    return new ReadWriteOptionsBuilder(null, DEFAULT_NUM_ATTEMPTS);
+  }
+
+  /**
+   * @return a new builder with default values applicable to configuring options for a read-only
+   *     transaction.
    */
   @Nonnull
-  public static TransactionOptions create(@Nonnull Executor executor, int numberOfAttempts) {
-    Preconditions.checkArgument(numberOfAttempts > 0, "You must allow at least one attempt");
-    return new TransactionOptions(numberOfAttempts, executor);
+  public static ReadOnlyOptionsBuilder createReadOnlyOptionsBuilder() {
+    return new ReadOnlyOptionsBuilder(null, null);
+  }
+
+  @InternalExtensionOnly
+  public abstract static class Builder<B extends Builder<B>> {
+    @Nullable protected Executor executor;
+
+    protected Builder(@Nullable Executor executor) {
+      this.executor = executor;
+    }
+
+    /**
+     * @return The {@link Executor} user callbacks will execute on, If null, the default executor
+     *     will be used.
+     */
+    @Nullable
+    public Executor getExecutor() {
+      return executor;
+    }
+
+    /**
+     * @param executor The {@link Executor} user callbacks will executed on. If null, the default
+     *     executor will be used.
+     * @return {@code this} builder
+     */
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    public B setExecutor(@Nullable Executor executor) {
+      this.executor = executor;
+      return (B) this;
+    }
+
+    /** @return an instance of {@link TransactionOptions} from the values passed to this builder */
+    @Nonnull
+    public abstract TransactionOptions build();
+  }
+
+  /**
+   * A typesafe builder class representing those options that are applicable when configuring a
+   * transaction to be read-only. All methods function as "set" rather than returning a new copy
+   * with a value set on it.
```

Comment:
I would drop "All methods function as "set" rather than returning a new copy with a value set on it." since this seems expected.